### PR TITLE
Fix many bugs

### PR DESCRIPTION
**Pull request by imkiva**
1. Fixed part of #381, now we can detect circular dependency like `def a a => xxxx`, but `def a a` is a more complicated problem, which will be addressed in another PR (possible solution #384).
2. Fixed library compiler's error reporting
3. The little typer should check the hole's solution, or the following code will crash the compiler
    ```
    def a => b
    def b => a
    ```
    when trying to solve metas generated for result type of `a` and `b`

### CODE REVIEW

**Comment by ice1000** [comment]
With my commit, `def a : Type` will be reported with an error, but `def a a` still triggers NPE.

File: base/src/main/java/org/aya/tyck/StmtTycker.java [hunk 1]
```diff
@@ -137,6 +138,8 @@ public void tyckHeader(@NotNull Decl decl, @NotNull ExprTycker tycker) {
         // It might contain unsolved holes, but that's acceptable.
         var resultRes = tycker.synthesize(fn.result).wellTyped().freezeHoles(tycker.state);
         fn.signature = new Def.Signature(resultTele, resultRes);
+        if (resultTele.isEmpty() && fn.body.isRight() && fn.body.getRightValue().isEmpty())
+          reporter.report(new NobodyError(decl.sourcePos, fn.ref));
```
Comment (imkiva) [inline base/src/main/java/org/aya/tyck/StmtTycker.java:142]: Can we do this check before tycking the telescope? So `def a a` would be rejected nicely.

Comment (ice1000) [reply base/src/main/java/org/aya/tyck/StmtTycker.java:142]: `def a a` does not have empty telescope.

Comment (ice1000) [reply base/src/main/java/org/aya/tyck/StmtTycker.java:142]: If your telescope uses yourself, you should reject the function

Comment (imkiva) [reply base/src/main/java/org/aya/tyck/StmtTycker.java:142]: Yes. My brain must be offline. 😢

**Comment by ice1000** [comment]
Fixed conflicts

**Review by ice1000 — APPROVED** [review]
review submitted as commit
bors r+

**Pull request merged**